### PR TITLE
Enabling multilingual models for translation pipelines.

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

Enables mutlilingual translation for pipelines.

Some models can target multiple languages/ language pairs. Before this PR, there was no simple way to exploit that within the Translation pipeline.

## Move away from `translation_XX_to_YY`.

Because src_lang, tgt_lang pairs can be used for a single model, we need to move away from this task naming scheme.
Currently it was done, as being able to use `src_lang` and `tgt_lang` both within `pipeline(.... src_lang=XX, tgt_lang=yy)` and at call time `translator = pipeline(...); translation("input string", src_lang=XX, tgt_lang=YY)`

 ## Rely on the model's tokenizer to build the inputs.

We now have at least 3 (m2m. mbart50, T5) different models that prepare input ids in a different manner. In order to avoid switches within a pipeline that would depend on a model, instead `tokenizer` can optionally implement `_build_translation_input_ids`.  That enables to put model specific logic within their model files and use all their custom methods over there.

This is in line with https://github.com/huggingface/transformers/pull/10002.

## Misc

- `ensure_tensor_on_device` now supports non tensor members
- Added a `deep_round` test utility to enable testing nested structures that contain tensors, floats and so on.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/mbart/tokenization_mbart50_fast.py [hunk 1]
```diff
@@ -273,6 +273,15 @@ def set_tgt_lang_special_tokens(self, tgt_lang: str) -> None:
             special_tokens=list(zip(prefix_tokens_str + suffix_tokens_str, self.prefix_tokens + self.suffix_tokens)),
         )
 
+    def _build_translation_inputs(self, raw_inputs, src_lang: Optional[str], tgt_lang: Optional[str], truncation):
+        if src_lang is None or tgt_lang is None:
```
Comment (patil-suraj) [inline src/transformers/models/mbart/tokenization_mbart50_fast.py:277]: we should add this method in all many to many tokenizes (`MBart50Tokenizer`, `M2M100Tokenizer`)

File: src/transformers/pipelines/text2text_generation.py [hunk 2]
```diff
@@ -50,6 +52,31 @@ def check_inputs(self, input_length: int, min_length: int, max_length: int):
...
+                " `args[0]`: {} have the wrong format. The should be either of type `str` or type `list`".format(
+                    args[0]
+                )
+            )
+        inputs = super()._parse_and_tokenize(*args, padding=padding, truncation=truncation)
+        # This is produced by tokenizers but is an invalid
+        # generate kwargs
+        if "token_type_ids" in inputs:
+            del inputs["token_type_ids"]
+        return inputs
```
Comment (patil-suraj) [inline src/transformers/pipelines/text2text_generation.py:78]: can we instead pass `retuen_tokens_types_ids=False` as kwarg to `_parse_and_tokenize` ?

**Review by patil-suraj — COMMENTED** [review]
Thanks for doing this. I left a couple of comments.

File: src/transformers/pipelines/text2text_generation.py (hunk 2 above)
Comment (Narsil) [reply src/transformers/pipelines/text2text_generation.py:78]: Is it on by default now ?

Comment (patil-suraj) [reply src/transformers/pipelines/text2text_generation.py:78]: It is on for BERT like models, but It shouldn't be for seq2seq models. What model was failing because of this?

File: src/transformers/models/mbart/tokenization_mbart50_fast.py (hunk 1 above)
Comment (Narsil) [reply src/transformers/models/mbart/tokenization_mbart50_fast.py]: Done.

File: src/transformers/pipelines/text2text_generation.py (hunk 2 above)
Comment (Narsil) [reply src/transformers/pipelines/text2text_generation.py:78]: The translation tests I added in this PR were faling

**Comment by Narsil** [comment]
@LysandreJik 

I added a method here `deep_round` to try and make test equality a bit sane.

torchTensor(...) == torch.Tensor(..) does not work (understably).
Any sort of float comparison is also flaky.

`deep_round` simply tries to make `assertEqual` work in a sane way for any sort of nested structure to make test comparisons simpler to read and write.

Here the tests just need to make sure that the actual output of `_build_translation_input_ids` is actually correct, writing it in that way make its much more readable IMO (than having to extract each element and call `allclose` on them). It actually allowed me to see that mbart has a different encoding scheme.

How do you feel about such a function to improve small sanity checks ?

Another route would be to create something like `assertAlmostEqual` that behaves similarly but I think it's a bit  less simple to reason about.

Finally we could stick to not using any such helper functions.

File: src/transformers/testing_utils.py [hunk 3]
```diff
@@ -1121,3 +1121,24 @@ def execute_subprocess_async(cmd, env=None, stdin=None, timeout=180, quiet=False
...
+    elif isinstance(obj, (dict, BatchEncoding)):
+        return {deep_round(k, decimals): deep_round(v, decimals) for k, v in obj.items()}
+    elif isinstance(obj, (str, int)):
+        return obj
+    elif is_torch_available() and isinstance(obj, torch.Tensor):
+        return deep_round(obj.tolist())
+    elif isinstance(obj, float):
+        return round(obj, decimals)
+    else:
+        raise Exception(f"Not supported: {type(obj)}")
```
Comment (LysandreJik) [inline src/transformers/testing_utils.py:1144]: I'm not opposed to have this method! I think it could be made a bit better, for example with TF support. 

Also, wouldn't it be simpler to slightly change it so that it is a deep `torch.allclose`? We're used to having `torch.allclose` across the library, and we have some intuition of what `atol` or `rtol` we want. I think it would both stay closer to other tests, and make your testing code simpler, to have such a method.

I.e., instead of 

```py
self.assertEqual(
            deep_round(inputs),
            {
                # en_XX, A, test, EOS
                "input_ids": [[128022, 58, 4183, 2]],
                "attention_mask": [[1, 1, 1, 1]],
                # ar_AR
                "forced_bos_token_id": 44,
            },
        )
```

you would have

```py
self.assertTrue(
            deep_allclose(inputs, {
                # en_XX, A, test, EOS
                "input_ids": [[128022, 58, 4183, 2]],
                "attention_mask": [[1, 1, 1, 1]],
                # ar_AR
                "forced_bos_token_id": 44,
            }, atol=xxx, rtol=xxx)
        )
```

I don't feel strongly though, your current method (given TF support) is good already.

Comment (Narsil) [reply src/transformers/testing_utils.py:1144]: The biggest difference between your version and my version is understanding the `diff` when there is a mismatch.

If input_ids is changed, but the rest of the entries are equal, then usually the unittest output is very readable.

assertTrue on the other hand is a pretty nasty "False is not true" message, which doesn't really help (and for larger structures it's sometimes hard to pinpoint where the difference happened.

Your solution is what I was referring to when mentionning the `assertAlmostEqual` option (where I think we need to fuse `assertTrue` and `deep_allclose` to get consistent error messages, yet be able to keep allclose like comparisons.

For this PR, I'll stick to just add TF support for `deep_round` as its much simpler and keeps readable diffs.

Comment (LysandreJik) [reply src/transformers/testing_utils.py]: Ok, this works for me. Thanks!

**Comment by LysandreJik** [comment]
When this PR is ready, could you complete the description of the PR? It would help to understand what we're reviewing, and would be better for the release notes and posterity. Thanks!

**Comment by Narsil** [comment]
Good catch ! Updated and completed !
I think this is ready to merge if you're ok with it.

**Comment by LysandreJik** [comment]
I'm asking @patrickvonplaten and @sgugger for review as they're more acquainted with mBART-like tokenizers and their review would be helpful.

File: src/transformers/testing_utils.py [hunk 4]
```diff
@@ -1121,3 +1124,26 @@ def execute_subprocess_async(cmd, env=None, stdin=None, timeout=180, quiet=False
         raise RuntimeError(f"'{cmd_str}' produced no output.")
 
     return result
+
+
+def deep_round(obj, decimals=3):
```
Comment (sgugger) [inline src/transformers/testing_utils.py:1129]: The name is badly chosen IMO since you are using it primarily to convert tensors/arrays in the test below. I'd just name it `listify_and_round` (or `nested_listify_and_round` or `recursive_listify_and_round` if you really want to highlight the fact it's recursive, "deep" does not imply that IMO)

File: src/transformers/pipelines/text2text_generation.py [hunk 5]
```diff
@@ -50,6 +52,31 @@ def check_inputs(self, input_length: int, min_length: int, max_length: int):
...
+            padding = False
+        else:
+            raise ValueError(
+                " `args[0]`: {} have the wrong format. The should be either of type `str` or type `list`".format(
+                    args[0]
+                )
+            )
+        inputs = super()._parse_and_tokenize(*args, padding=padding, truncation=truncation)
+        # This is produced by tokenizers but is an invalid
+        # generate kwargs
```
Comment (sgugger) [inline src/transformers/pipelines/text2text_generation.py:75]:
```suggestion
        # This is produced by tokenizers but is an invalid generate kwargs
```
Nit: let's take advantage of the whole 119-chars line.

File: src/transformers/models/m2m_100/tokenization_m2m_100.py [hunk 6]
```diff
@@ -290,6 +290,15 @@ def prepare_seq2seq_batch(
         self.set_src_lang_special_tokens(self.src_lang)
         return super().prepare_seq2seq_batch(src_texts, tgt_texts, **kwargs)
 
+    def _build_translation_inputs(self, raw_inputs, src_lang: Optional[str], tgt_lang: Optional[str], **extra_kwargs):
+        if src_lang is None or tgt_lang is None:
```
Comment (sgugger) [inline src/transformers/models/m2m_100/tokenization_m2m_100.py:294]: A one-line docstring or comment for us internally to remember where this is used would be super helpful! (putting it on this one but it applies to the others as well).

**Review by sgugger — APPROVED** [review]
Thanks for adding this! Just have some cosmetic comments.

File: src/transformers/testing_utils.py (hunk 4 above)
Comment (Narsil) [reply src/transformers/testing_utils.py:1129]: `deep` does imply that for me (referring to `deepcopy`: https://docs.python.org/3/library/copy.html). But if it does not for you, then `nested` is fine by me.

I don't really like `listify` as it's not part of the intent, the intent is really simplifying an object for comparison. It just happens that tensors don't compare nicely (even if rounded) so I'm changing their type as a side effect (if tensors started to allow comparison, I would stop changing their type).

- `nested_round`
- `nested_simplify`

Wdyt ?

File: src/transformers/models/m2m_100/tokenization_m2m_100.py (hunk 6 above)
Comment (Narsil) [reply src/transformers/models/m2m_100/tokenization_m2m_100.py:294]: Done

File: src/transformers/testing_utils.py (hunk 4 above)
Comment (sgugger) [reply src/transformers/testing_utils.py:1129]: Let's go for `nested_simplify` then!

File: src/transformers/pipelines/text2text_generation.py [hunk 7]
```diff
@@ -230,6 +242,22 @@ class TranslationPipeline(Text2TextGenerationPipeline):
...
+    def __init__(self, *args, src_lang=None, tgt_lang=None, **kwargs):
+        super().__init__(*args, **kwargs)
+        if src_lang is not None:
+            self.src_lang = src_lang
+        if tgt_lang is not None:
+            self.tgt_lang = tgt_lang
+        if src_lang is None and tgt_lang is None:
+            task = kwargs.get("task", "")
+            tokens = task.split("_")
+            if task and len(tokens) == 4:
```
Comment (patrickvonplaten) [inline src/transformers/pipelines/text2text_generation.py:257]: maybe a else `raise ValueError(....)`

File: src/transformers/pipelines/text2text_generation.py [hunk 8]
```diff
@@ -230,6 +242,22 @@ class TranslationPipeline(Text2TextGenerationPipeline):
...
+
+    def __init__(self, *args, src_lang=None, tgt_lang=None, **kwargs):
+        super().__init__(*args, **kwargs)
+        if src_lang is not None:
+            self.src_lang = src_lang
+        if tgt_lang is not None:
+            self.tgt_lang = tgt_lang
+        if src_lang is None and tgt_lang is None:
+            task = kwargs.get("task", "")
+            tokens = task.split("_")
```
Comment (patrickvonplaten) [inline src/transformers/pipelines/text2text_generation.py:256]: nit - don't think `tokens` is a good name here -> `words` maybe or `items` => tokens makes me think about the tokenizers which are not related here at all IMO

**Review by patrickvonplaten — APPROVED** [review]
Thanks!

File: src/transformers/pipelines/text2text_generation.py (hunk 7 above)
Comment (Narsil) [reply src/transformers/pipelines/text2text_generation.py:257]: I don't think we should treat as `invalid` any other task argument here.
We are already within the `TranslationPipeline` so the actual task is  implied no matter the argument.
We are merely trying to backport `task=translation_xx_to_yy` when it's easy (.

But if code was written with `task=translation_en_XX_to_ar_AR` for instance, I don't want to hard fail either. Code will continue to work as it was before, and happily ignore the `src_lang` and `tgt_lang` here.

Note: The validation already exists for `pipeline` function : https://github.com/huggingface/transformers/blob/master/src/transformers/pipelines/__init__.py#L234

Note: Your comment led me to do some testing, and I found a bug within my implementation for m2m100 tokenizer.. Thanks.

Be

File: src/transformers/pipelines/text2text_generation.py (hunk 8 above)
Comment (Narsil) [reply src/transformers/pipelines/text2text_generation.py]: Fair enough

File: src/transformers/testing_utils.py (hunk 4 above)
Comment (Narsil) [reply src/transformers/testing_utils.py]: Done !
